### PR TITLE
fix(vllm): include image geometry in multimodal hash preimage

### PR DESCRIPTION
**Pull request by nnshah1**
## Summary
- Harden `compute_mm_uuids_from_images` so RGB images with different `(W, H)` but equal pixel count produce distinct UUIDs. The preimage is now a fixed-length 12-byte header `(version || mode || dtype || channels || H || W)` prefixed to the pixel bytes and fed to blake3 incrementally.
- Tighten the input contract: PIL must be `mode="RGB"`, ndarray must be `uint8` with shape `(H, W, 3)`, non-contiguous ndarrays are coerced via `np.ascontiguousarray`, everything else raises. The unused `bytes` passthrough is removed.
- Add `components/src/dynamo/vllm/tests/multimodal_utils/test_vllm_hash_utils.py` covering dimension-swap regression (PIL + ndarray), PIL/ndarray parity, contract rejections, non-contiguous coercion, and a pinned-digest stability anchor.

## Test plan
- [x] `pytest components/src/dynamo/vllm/tests/multimodal_utils/test_vllm_hash_utils.py` — 11 pass
- [x] `pytest components/src/dynamo/vllm/tests/` — 338 pass, 9 skipped (pre-existing, unrelated)
- [x] `black --check --target-version=py312` clean on both changed files
- [x] Latency sanity validated offline at 64×64 / 512×512 / 2048×2048 RGB — hardened path stays within ~1.5× of a raw `blake3(arr.tobytes())` baseline

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

<a href="https://nvidia.devinenterprise.com/review/ai-dynamo/dynamo/pull/8341" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved image identification in multimodal processing by incorporating image dimensions into hash computation, preventing collisions between images with identical pixels but different sizes.
  * Strengthened image input validation with stricter format requirements.

* **Tests**
  * Added comprehensive test coverage for image hashing behavior and edge cases.

### CODE REVIEW

File: components/src/dynamo/vllm/multimodal_utils/hash_utils.py [hunk 1]
```diff
@@ -2,33 +2,88 @@
...
     uuids: list[str] = []
     for img in images:
-        raw_bytes = image_to_bytes(img)
-        uuids.append(blake3.blake3(raw_bytes).hexdigest())
+        header, pixels = _image_preimage_parts(img)
+        h = blake3.blake3()
+        h.update(header)
+        h.update(pixels)
+        uuids.append(h.hexdigest())
     return uuids
```
[the remark this answers is not in the document]
Comment (nnshah1) [reply components/src/dynamo/vllm/multimodal_utils/hash_utils.py:88]: @rmccorm4 - I think this is ok - but wanted to double check

File: components/src/dynamo/vllm/multimodal_utils/hash_utils.py [hunk 2]
```diff
@@ -2,33 +2,88 @@
...
+    Raises:
+        ValueError: input shape, dtype, or mode violates the RGB uint8 contract.
+        TypeError: input is neither a PIL.Image.Image nor an np.ndarray.
+    """
     from PIL import Image
 
-    if isinstance(img, bytes):
-        return img
+    if isinstance(img, Image.Image):
+        if img.mode != "RGB":
```
Comment (nnshah1) [inline components/src/dynamo/vllm/multimodal_utils/hash_utils.py:50]: question: @rmccorm4, @GuanLuo  is enforcing RGB mode here too strict?

File: components/src/dynamo/vllm/tests/multimodal_utils/test_vllm_hash_utils.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+the on-disk preimage format via a known-digest stability anchor.
+"""
+
+import numpy as np
+import pytest
+from PIL import Image
+
+from dynamo.vllm.multimodal_utils.hash_utils import compute_mm_uuids_from_images
+
+pytestmark = [pytest.mark.pre_merge, pytest.mark.vllm]
```
[the remark this answers is not in the document]
Comment (nnshah1) [reply components/src/dynamo/vllm/tests/multimodal_utils/test_vllm_hash_utils.py]: Fixed in cd121867f3 — pytestmark now matches the sibling test_vllm_embedding_loader.py exactly: [pre_merge, vllm, gpu_0, multimodal].

File: components/src/dynamo/vllm/multimodal_utils/hash_utils.py [hunk 4]
```diff
@@ -2,33 +2,88 @@
...
+
+    The returned pair is safe to feed into an incremental blake3 hasher. Two
+    images that differ only in (H, W) produce different `header_bytes` and
+    therefore different digests.
+
+    Raises:
+        ValueError: input shape, dtype, or mode violates the RGB uint8 contract.
+        TypeError: input is neither a PIL.Image.Image nor an np.ndarray.
+    """
     from PIL import Image
```
[the remark this answers is not in the document]
Comment (nnshah1) [reply components/src/dynamo/vllm/multimodal_utils/hash_utils.py]: Fixed in 01e8d16a3e — from PIL import Image moved to module top alongside the other imports.

[the remark this answers is not in the document]
Comment (nnshah1) [reply components/src/dynamo/vllm/multimodal_utils/hash_utils.py]: Fixed in 01e8d16a3e — from PIL import Image moved to module top alongside the other imports.

File: components/src/dynamo/vllm/multimodal_utils/hash_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nnshah1) [reply components/src/dynamo/vllm/multimodal_utils/hash_utils.py:88]: Intentional; the PR body covers this: cache entries live in-memory per worker, so the new preimage format naturally evicts stale entries on rolling restart. A reserved version byte lets the preimage format evolve without silent cross-version overlap. No migration path needed.

[the remark this answers is not in the document]
Comment (nnshah1) [reply components/src/dynamo/vllm/multimodal_utils/hash_utils.py:88]: Fixed in 01e8d16a3e — the _compute_mm_uuids docstring in handlers.py now references blake3 and the fixed-length header + pixel preimage.

File: components/src/dynamo/vllm/multimodal_utils/hash_utils.py [hunk 5]
```diff
@@ -2,33 +2,87 @@
...
+#     0      1     version_byte   currently 0x01; bump to rotate preimage format
+#     1      1     mode_tag       0x00 = RGB
+#     2      1     dtype_tag      0x00 = uint8
+#     3      1     channels       currently 3
+#     4      4     height         u32 little-endian
+#     8      4     width          u32 little-endian
+#
+# The fixed-length header prevents delimiter-ambiguity collisions: no two
+# distinct (header, pixel) pairs can produce the same preimage.
+_HEADER_STRUCT = struct.Struct("<BBBBII")
```
Comment (grahamking) [inline components/src/dynamo/vllm/multimodal_utils/hash_utils.py:32]: What does "<BBBBII" mean? I'm guessing a format definition, Boolean and Integer? Could you add a comment?

File: components/src/dynamo/vllm/multimodal_utils/hash_utils.py [hunk 6]
```diff
@@ -2,33 +2,87 @@
...
+def _header(height: int, width: int) -> bytes:
+    return _HEADER_STRUCT.pack(
+        _VERSION_BYTE, _MODE_RGB, _DTYPE_UINT8, _CHANNELS_RGB, height, width
+    )
 
-    if isinstance(img, Image.Image | np.ndarray):
-        return img.tobytes()
 
-    raise TypeError(f"Unsupported image type for hashing: {type(img)}")
+def _image_preimage_parts(img: Any) -> tuple[bytes, bytes]:
```
Comment (grahamking) [inline components/src/dynamo/vllm/multimodal_utils/hash_utils.py]: ~Should this be `img: image.Image`.~

Never mind, the indenting is hard to read in the github diff viewer.

Could we replace this `Any` with the two types, so we get static checking?

File: components/src/dynamo/vllm/tests/multimodal_utils/test_vllm_hash_utils.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (grahamking) [inline components/src/dynamo/vllm/tests/multimodal_utils/test_vllm_hash_utils.py:1]: @nnshah1 Could you go through these tests and see what the minimum necessary set is? Agents often add too many tests.

File: components/src/dynamo/vllm/multimodal_utils/hash_utils.py (hunk 5 above)
Comment (nnshah1) [reply components/src/dynamo/vllm/multimodal_utils/hash_utils.py:32]: Added in 01a58088f5 — split the format chars into their own block explaining `<` (little-endian, no padding), `BBBB` (four uint8 fields: version, mode, dtype, channels), and `II` (two uint32 fields: height, width).

File: components/src/dynamo/vllm/multimodal_utils/hash_utils.py (hunk 6 above)
Comment (nnshah1) [reply components/src/dynamo/vllm/multimodal_utils/hash_utils.py]: Done in 0d5fbc9c2c — introduced an `ImageInput = Union[Image.Image, np.ndarray]` alias and threaded it through `_image_preimage_parts()` and `compute_mm_uuids_from_images()`. `Any` is gone from this module.

File: components/src/dynamo/vllm/tests/multimodal_utils/test_vllm_hash_utils.py (hunk 7 above)
Comment (nnshah1) [reply components/src/dynamo/vllm/tests/multimodal_utils/test_vllm_hash_utils.py:1]: Good call — trimmed in af2d3fdfc2, 11 → 7 tests. Dropped:

- `test_pil_ndarray_equivalent_inputs_match` — implied by the dimension-swap parametrization + known-digest anchor.
- `test_non_contiguous_ndarray_coerced` — `ndarray.tobytes()` normalizes strides in C order, so this was testing numpy, not us.
- `test_rejects_invalid_input[pil_mode_RGBA]` and `[ndarray_shape_2d]` — same branches as the `pil_mode_L` and `ndarray_shape_4ch` cases already kept.

Kept: 2× dimension-swap (PIL + ndarray), 4× invalid-input rejection (one per distinct branch: PIL mode, ndarray dtype, ndarray channels, non-image type), and the known-digest stability anchor.

**Comment by furionw** [comment]
I had [#8340](https://github.com/ai-dynamo/dynamo/pull/8340) for the same issue, while I like your solution better. I have no objection of landing this PR

[both embedding cache and mm routing is using this code](https://github.com/search?q=repo%3Aai-dynamo%2Fdynamo%20compute_mm_uuids_from_images&type=code). The changes seems reasonable from multimodal point of view
- embedding cache is not persisted
- mm routing
  - [compatability] KV cache of multimodal is not persisted to disk either (KVBM doesn't support multimdaol yet)
  - [functionality] both router and worker are using the same function. so UUID is aligned

File: components/src/dynamo/vllm/multimodal_utils/hash_utils.py (hunk 2 above)
Comment (furionw) [reply components/src/dynamo/vllm/multimodal_utils/hash_utils.py:57]: our ImageLoader always convert Image to RGB. should be fine for now

```
    @staticmethod
    def _open_image_sync(image_data: BytesIO) -> Image.Image:
        """Open, validate, and decode an image from raw bytes. Runs in a thread."""
        image = Image.open(image_data, formats=["JPEG", "PNG", "WEBP"])
        if image.format not in ("JPEG", "PNG", "WEBP"):
            raise ValueError(f"Unsupported image format: {image.format}")
        # Image.open() is lazy — convert() forces the actual pixel decode
        return image.convert("RGB")
```

I thought of cases like

  - vanilla dynamo with http data
  - vanilla dynamo with base64
  - frontend decoding
  - disagg encoder
  - aggregated node
  - pre-computed mm embeddings
  
they should all work with this change (some of them are orthogonal to this change)

**Review by furionw — APPROVED** [review]
Thank you!
